frontend/components/PaginatedList/PaginatedList.tsx

```diff
@@ -0,0 +1,189 @@
+          setError(err as Error);
+        }
+      } finally {
+        if (!isCancelled) {
+          setIsLoading(false);
+        }
+      }
+    }
+
+    loadPage();
+
+    return () => {
+      isCancelled = true;
+    };
+  }, [currentPage, fetchPage]);
+
+  // Create an imperative handle for this component so that parents
+  // can call `ref.current.getDirtyItems()` to get the changed set.
+  useImperativeHandle(ref, () => ({
+    getDirtyItems() {
+      return Object.values(dirtyItems);
+    },
+  }));
+
+  // TODO -- better error state?
+  if (error) return <p>Error: {error.message}</p>;
+
+  // Redner the list.
+  return (
+    <div className={baseClass}>
+      {isLoading && (
+        <div className="loading-overlay">
+          <Spinner />
+        </div>
+      )}
+      <ul className={`${baseClass}__list`}>
+        {items.map((_item) => {
+          // If an item has been marked as changed, use the changed version
+          // of the item rather than the one from the page fetch.  This allows
+          // us to render an item correctly even after we've navigated away
+          // from its page and then back again.
+          const item = dirtyItems[_item[idKey]] ?? _item;
+          return (
+            <li className={`${baseClass}__row`} key={item[idKey]}>
+              <Checkbox
+                value={
+                  typeof isSelected === "function"
+                    ? isSelected(item)
+                    : item[isSelected]
+                }
+                name={`item_${item[idKey]}_checkbox`}
+                onChange={() => {
+                  // When checkbox is toggled, set item as dirty.
+                  // The parent is responsible for actually updating item properties via onToggleItem().
+                  setDirtyItems({
+                    ...dirtyItems,
+                    [item[idKey]]: onToggleItem(item),
+                  });
+                }}
+              >
+                {renderItemLabel ? (
+                  renderItemLabel(item)
+                ) : (
+                  <span>{item[labelKey]}</span>
+                )}
+              </Checkbox>
+              {renderItemRow &&
+                // If a custom row renderer was supplied, call it with the item value
+                // as well as the callback the parent can use to indicate changes to an item.
+                renderItemRow(item, (changedItem) => {
+                  setDirtyItems({
+                    ...dirtyItems,
+                    [changedItem[idKey]]: changedItem,
+                  });
+                })}
+            </li>
+          );
+        })}
+      </ul>
+      <Pagination
+        resultsOnCurrentPage={items.length}
+        currentPage={currentPage}
+        resultsPerPage={pageSize}
+        onPaginationChange={setCurrentPage}
+      />
+    </div>
+  );
+}
+
+// Wrap with forwardRef to expose the imperative handle:
+const PaginatedList = forwardRef(PaginatedListInner) as <TItem>(
```

Comment:
It looks like `forwardRef` will be [deprecated soon](https://react.dev/reference/react/forwardRef),
is there a way to do this with the `ref` passed in as a prop instead?

Reply:
Ah I see...until we're on v19, you need to use `forwardRef` for the `useImperativeHandle` above. Can
we just leave a note then to change this whenever we upgrade to v19?